session/src/main/java/org/apache/iotdb/session/template/Template.java

```diff
@@ -0,0 +1,241 @@
+  }
+
+  public List<Boolean> getAlignedList() {
+    return this.alignedList;
+  }
+
+  public List<String> getMeasurementsPath() {
+    return this.measurementsPath;
+  }
+
+  public List<TSEncoding> getEncodings() {
+    return encodings;
+  }
+
+  public List<CompressionType> getCompressors() {
+    return compressors;
+  }
+
+  // region Interface to manipulate Template
+
+  public void addToTemplate(Node child) throws StatementExecutionException {
+    if (moutedNode.getChildren().containsKey(child.getName()))
+      throw new StatementExecutionException("Duplicated child of node in template.");
+    moutedNode.getChildren().put(child.getName(), child);
+  }
+
+  public void deleteFromTemplate(String name) throws StatementExecutionException {
+    if (moutedNode.getChildren().containsKey(name)) moutedNode.getChildren().remove(name);
+    else throw new StatementExecutionException("It is not a direct child of the template: " + name);
+  }
+
+  public void traverseTemplate() throws StatementExecutionException {
+    List<String> result = new ArrayList<>();
+    List<String> nowPaths = new ArrayList<>();
+    Deque<Integer> childSize = new ArrayDeque<>();
+    Deque<Boolean> alignStack = new ArrayDeque<>();
+    Deque<Node> stack = new ArrayDeque<>();
+    this.measurementsPath.clear();
+    this.alignedList.clear();
+    this.dataTypes.clear();
+    this.encodings.clear();
+    this.compressors.clear();
+
+    Node cur;
+    MeasurementNode mNode;
+    int count;
+    boolean isAligned;
+
+    for (Node child : moutedNode.getChildren().values()) stack.push(child);
+    count = moutedNode.getChildren().size();
+    isAligned = moutedNode.isShareTime();
+
+    while (stack.size() != 0) {
+      if (stack.size() > 64)
+        throw new StatementExecutionException(
+            "Too long prefix in template, please check iterations.");
+
+      cur = stack.pop();
+      count--;
+
+      if ((count < 0) && (childSize.size() > 0)) {
+        isAligned = alignStack.pop();
+        count = childSize.pop();
+        count--;
+        nowPaths.remove(nowPaths.size() - 1);
+      }
+
+      if (cur != moutedNode) nowPaths.add(cur.getName());
+
+      if (cur.getChildren() != null) {
+        childSize.push(count);
+        alignStack.push(isAligned);
+
+        for (Node child : cur.getChildren().values()) stack.push(child);
+        count = cur.getChildren().size();
+        isAligned = cur.isShareTime();
+      }
+
+      if (cur.isMeasurement()) {
+        mNode = (MeasurementNode) cur;
+        result.add(convertList2String(nowPaths));
+        measurementsPath.add(convertList2String(nowPaths));
+        alignedList.add(isAligned);
+        dataTypes.add(mNode.getDataType());
+        encodings.add(mNode.getEncoding());
+        compressors.add(mNode.getCompressionType());
+        nowPaths.remove(nowPaths.size() - 1);
+      }
+    }
+    measurementsPath = result;
+  }
+
+  public byte[] serialize() throws IOException {
```

Comment:
```suggestion
  public byte[] serialize(ByteBuffer or OutputStream) throws IOException {
```

refer to  serializeTo() in ChunkMetadata

Reply:
change serialize to :
public void serialize(OutputStream)